Add option for truncated stream detection

mfkl opened the pull request:
Having another go at https://github.com/dotnet/runtime/issues/47563 (previous https://github.com/dotnet/runtime/pull/61768)

This time using an opt-in flag to avoid breakage like this https://github.com/dotnet/runtime/issues/72726.

/cc @Jozkee

mfkl committed:
Add option for truncated stream detection

fix https://github.com/dotnet/runtime/issues/47563

stephentoub commented on src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs:
```diff
@@ -479,6 +478,8 @@ async Task<long> GetLengthAsync(CompressionLevel compressionLevel)
         [InlineData(TestScenario.ReadByteAsync)]
         public async Task StreamTruncation_IsDetected(TestScenario scenario)
         {
+            AppContext.SetSwitch("System.IO.Compression.UseStrictValidation", true);
```
This affects all tests process-wide. Any test that changes such a process-wide switch needs to use RemoteExecutor to move that code out of process.

mfkl replied:
Ok, done.

Jozkee replied:
Another option is to create a second test project that enables the appcontext switch and runs all tests of the "normal" project with the switch on.

Jozkee replied:
And a third option could be to turn the switch on for all the tests, we already know our tests will pass with the "strict validation" in place.

mfkl replied:
> a third option could be to turn the switch on for all the tests, we already know our tests will pass with the "strict validation" in place.

Not anymore, this one will fail https://github.com/dotnet/runtime/blob/eead714e4a5e82f2075de7d5eaad75dedf509941/src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Deflate.cs#L78

> Another option is to create a second test project that enables the appcontext switch and runs all tests of the "normal" project with the switch on.

I prefer the current way (with the RemoteExecutor) but no strong feelings. Let me know.

stephentoub commented on src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs:
```diff
@@ -146,6 +151,42 @@ public static void StreamsEqual(Stream ast, Stream bst, int blocksToRead)
+                    bd = NormalizeLineEndings(bd);
+                }
+
+                Assert.True(ArraysEqual<byte>(ad, bd, ac), "Stream contents not equal: " + ast.ToString() + ", " + bst.ToString());
```
This should use AssertExtensions.SequenceEqual.

mfkl replied:
Done

stephentoub commented on src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs:
```diff
@@ -227,9 +234,23 @@ private bool TryDecompress(Span<byte> destination, out int bytesWritten)
+
+            string? envVar = Environment.GetEnvironmentVariable("DOTNET_SYSTEM_COMPRESSION_STRICT_VALIDATION");
+            return envVar is not null && (envVar == "1" || envVar.Equals("true", StringComparison.OrdinalIgnoreCase));
+        }
```
This should be cached into a static readonly bool.

mfkl replied:
I considered that but thought we wanted to be able to change that during runtime. Guess not. Changed.

stephentoub commented on src/libraries/System.IO.Compression/src/Resources/Strings.resx:
```diff
@@ -278,6 +278,9 @@
     <value>Split or spanned archives are not supported.</value>
   </data>
+  <data name="TruncatedData" xml:space="preserve">
+    <value>Found truncated data while decoding.</value>
```
The text of the same error message should be the same across libs.

stephentoub commented on src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:
```diff
@@ -1049,5 +1078,16 @@ private void AsyncOperationStarting()
+
+            string? envVar = Environment.GetEnvironmentVariable("DOTNET_SYSTEM_COMPRESSION_STRICT_VALIDATION");
+            return envVar is not null && (envVar == "1" || envVar.Equals("true", StringComparison.OrdinalIgnoreCase));
+        }
```
As in the other library, this needs to be cached into a static readonly bool.

mfkl replied:
Yes, changed.

mfkl committed:
Use RemoteExecutor

move the test to concrete classes as abstracted classes are not supported by RemoteExecutor

mfkl committed:
review feedback

mfkl committed:
use same error text message

mfkl committed:
cache appcontext getswitch

mfkl commented on src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs:
```diff
@@ -146,6 +151,42 @@ public static void StreamsEqual(Stream ast, Stream bst, int blocksToRead)
+                    break;
+
+                ac = await ReadAllBytesAsync(ast, ad, 0, 4096);
+                bc = await ReadAllBytesAsync(bst, bd, 0, 4096);
```
Ok

stephentoub commented on src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs:
```diff
@@ -146,6 +151,42 @@ public static void StreamsEqual(Stream ast, Stream bst, int blocksToRead)
+                    break;
+
+                ac = await ast.ReadAtLeastAsync(ad, 4096);
+                bc = await bst.ReadAtLeastAsync(bd, 4096);
```
Presumably you want to pass throwOnEndOfStream:false for both of these?

mfkl replied:
Yes.. thanks.

stephentoub commented on src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs:
```diff
@@ -146,6 +151,42 @@ public static void StreamsEqual(Stream ast, Stream bst, int blocksToRead)
+                    bd = NormalizeLineEndings(bd);
+                }
+
+                AssertExtensions.SequenceEqual(ad, bd);
```
And these should be sliced, e.g. `ad.AsSpan(0, ac), bd.AsSpan(0, bc)`.

If this is just copying what's already there for the sync implementation, then maybe inputs just aren't triggering the faulty behavior?

stephentoub commented on src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs:
```diff
@@ -227,9 +234,15 @@ private bool TryDecompress(Span<byte> destination, out int bytesWritten)
             return false;
         }
 
+        private static readonly bool UseStrictValidation =
```
UseStrictValidation => s_useStrictValidation

mfkl replied:
Ok

mfkl committed:
fix failing test

mfkl committed:
slice byte array for assertion

mfkl committed:
renaming

mfkl committed:
add missing RemoteExecutor.IsSupported

mfkl commented:
CI is fixed. 

Feel free to let me know what you prefer regarding https://github.com/dotnet/runtime/pull/75671#discussion_r977370461 and if anyone has any other feedback regarding this PR.

Jozkee commented on src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.ZLib.cs:
```diff
@@ -19,44 +21,134 @@ public class ZLibStreamUnitTests : CompressionStreamUnitTestBase
+        [InlineData(TestScenario.ReadByte)]
+        [InlineData(TestScenario.ReadByteAsync)]
+        [ConditionalTheory(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
+        public void StreamTruncation_IsDetected(TestScenario testScenario)
```
This test is duplicated for Deflate, Gzip and Zlib tests. Why did you move it out of `CompressionStreamUnitTestBase`?

mfkl replied:
https://github.com/dotnet/runtime/pull/75671/commits/fa4b613a9c3fe4f13d8b2fc147676ca2d2e92236

> abstracted classes are not supported by RemoteExecutor

Jozkee commented on src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs:
```diff
@@ -275,7 +277,6 @@ public void DerivedStream_ReadWriteSpan_UsesReadWriteArray()
 
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/47563")]
         [Fact]
         public void StreamCorruption_IsDetected()
```
This test also needed the RemoteExecutor + enable the switch, no? I am a bit surprised that CI didn't fail.

mfkl replied:
Gzip detects the corruption without the strict validation enabled (thanks to its CRC integrity check, I assume). 

zlib not so much, if I remove the `AppContext.SetSwitch("System.IO.Compression.UseStrictValidation", true)`, the test fails.

mfkl committed:
fast check first

mfkl commented:
I believe I've addressed most concerns. Do you have any more feedback on this PR?

jozkee reviewed (approved):
LGTM, thanks.

CI issue looks unrelated to this PR.
